### ISSUE TITLE
Design default testing functions for log priors, log likelihoods and toy distributions

### ISSUE THREAD

**Issue by ben18785** [pints-team/pints#743]
A lot of tests we need to implement are shared across different objects.

For example, for a log prior, we want to test:

- instantiation
- that it can be called on a list object to produce an output
- mean equal to its theoretical value
- correct number of parameters
- samples of the correct shape
- it raise value errors if called on an incorrectly shaped object
- that it outputs a sensitivity when evaluateS1 is called

Having a testing function that _all_ log priors are required to be tested with would ensure that we get consistent implementations and would have caught some of the subtle implementational differences I've found for some functions. These aren't supposed to do everything - for example, value based tests on outputs aren't covered - but they would catch a lot of things and would ensure that we test things in a consistent manner.

Whilst we may not be at the stage yet where it makes sense to build testing functions, I do think we may want to consider these in the future, especially as others start to contribute to PINTS.

**Comment by MichaelClerx** [comment]
Agreed! It's fairly simple to write something that loops over all existing LogPriors in pints and performs some kind of test. (Bit trickier making the test output useful, as "error on line 8" no longer points you to which logprior it was that failed)

**Comment by MichaelClerx** [comment]
Alternatively maybe a shared LogPriorTestCase or something they can extend?
